### PR TITLE
Refactor custom direct methodology

### PR DESCRIPTION
**Pull request by riemanli**
## TL;DR
This pull request introduces changes to the `direct_computation.proto` and `protocol_config.proto` files. The changes include the addition of a new message `FrequencyVariances` and a `oneof` field `variance` in the `CustomDirectMethodology` message. It also modifies the comment for `custom_direct_methodology` in `ProtocolConfig`.
> 
## What changed
In `direct_computation.proto`, the `variance` field in `CustomDirectMethodology` has been replaced with a `oneof` field that can either be a `scalar` or `FrequencyVariances`. The `FrequencyVariances` message includes two maps for `variances` and `k_plus_variances`, which represent the variances of different types of frequency distributions from frequency 1 to the maximum frequency specified in the measurement spec.

> 
In `protocol_config.proto`, the comment for `custom_direct_methodology` has been updated to include a requirement for data providers to guarantee the independence of the result by salting their hash functions.
> 
## Why make this change
This change allows for more flexibility in specifying the variance in the `CustomDirectMethodology` message. It also clarifies the requirements for data providers who choose to use a custom direct methodology.

### CODE REVIEW

**Comment by wfa-reviewable** [comment]
This change is [Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/178)

**Comment by riemanli** [comment]
Current dependencies on/for this PR:
* main
  * **PR #178** <a href="https://app.graphite.dev/github/pr/world-federation-of-advertisers/cross-media-measurement-api/178" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/world-federation-of-advertisers/cross-media-measurement-api/178?utm_source=stack-comment).

**Review by SanjayVas — APPROVED** [review]
> Reviewed 1 of 1 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/178)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @kungfucraig)

**Review by jiayu-google — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/178)* status: 0 of 2 files reviewed, 2 unresolved discussions (waiting on @kungfucraig, @riemanli, @SanjayVas, and @stevenwarejones)

___
*[`src/main/proto/wfa/measurement/api/v2alpha/protocol_config.proto` line 144 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/178#-NfSSFqL4KNxn36ARov2:-NfSSFqL4KNxn36ARov3:bh4qvxn) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/311d0595d7bd2d3956232a960d6dac358ef1ea9c/src/main/proto/wfa/measurement/api/v2alpha/protocol_config.proto#L144)):*
> ```proto
>     // Used when data provider wants to use a methodology that is not listed in
>     // direct protocol config to compute direct measurements. Data providers who
>     // use this methodology must guarantee the independence of the result by
> ```

nit: use this methodology -> use any custom direct methodology
___
*[`src/main/proto/wfa/measurement/api/v2alpha/protocol_config.proto` line 145 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/178#-NfSRyAnELI_eETN54X2:-NfSRyAoBoAwBKPB65r7:b-snglb8) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/311d0595d7bd2d3956232a960d6dac358ef1ea9c/src/main/proto/wfa/measurement/api/v2alpha/protocol_config.proto#L145)):*
> ```proto
>     // direct protocol config to compute direct measurements. Data providers who
>     // use this methodology must guarantee the independence of the result by
>     // salting their hash functions.
> ```

Please note that data providers should never salt their hash functions when creating a sketch for MPC measurements.

**Review by kungfucraig — CHANGES_REQUESTED** [review]
> Reviewed 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/178)* status: all files reviewed, 3 unresolved discussions (waiting on @jiayu-google, @riemanli, @SanjayVas, and @stevenwarejones)

___
*[`src/main/proto/wfa/measurement/api/v2alpha/direct_computation.proto` line 42 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/178#-NfSVosB1DiPIP0tJdyX:-NfSVosB1DiPIP0tJdyY:bnuuhip) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/311d0595d7bd2d3956232a960d6dac358ef1ea9c/src/main/proto/wfa/measurement/api/v2alpha/direct_computation.proto#L42)):*
> ```proto
>   oneof variance {
>     // The variance when the result is a scalar type.
>     double scalar = 1;
> ```

I don't think we should rename this field as it will break anyone who migrates to it before the next release.
___
*[`src/main/proto/wfa/measurement/api/v2alpha/protocol_config.proto` line 145 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/178#-NfSRyAnELI_eETN54X2:-NfSW1K_CN47PAlx4Jcu:b-mwoy09) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/311d0595d7bd2d3956232a960d6dac358ef1ea9c/src/main/proto/wfa/measurement/api/v2alpha/protocol_config.proto#L145)):*
<details><summary><i>Previously, jiayu-google wrote…</i></summary><blockquote>

Please note that data providers should never salt their hash functions when creating a sketch for MPC measurements.
</blockquote></details>

Can we say "This guidance only applies to direct measurements, not MPC measurements."

I think eventually I'd like the Kingdom to generate a distinct salt for each MPC measurement. This would help ensure that if anyone got a hold of raw MPC data the ability to use it and combine it with other data would be limited. So, mostly just objecting to including the word "never."

**Comment by riemanli** [comment]
*[`src/main/proto/wfa/measurement/api/v2alpha/direct_computation.proto` line 42 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/178#-NfSVosB1DiPIP0tJdyX:-NfSZ47v3UbMeEpqFkbA:b-lbwi55) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/311d0595d7bd2d3956232a960d6dac358ef1ea9c/src/main/proto/wfa/measurement/api/v2alpha/direct_computation.proto#L42)):*
<details><summary><i>Previously, kungfucraig (Craig Wright) wrote…</i></summary><blockquote>

I don't think we should rename this field as it will break anyone who migrates to it before the next release.
</blockquote></details>

Could we ask them not to migrate the custom direct methodology before the next release?

**Review by kungfucraig — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/178)* status: all files reviewed, 3 unresolved discussions (waiting on @jiayu-google, @riemanli, @SanjayVas, and @stevenwarejones)

___
*[`src/main/proto/wfa/measurement/api/v2alpha/direct_computation.proto` line 42 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/178#-NfSVosB1DiPIP0tJdyX:-NfSbHVX1FtrfE2TeuEn:bn663s8) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/311d0595d7bd2d3956232a960d6dac358ef1ea9c/src/main/proto/wfa/measurement/api/v2alpha/direct_computation.proto#L42)):*
<details><summary><i>Previously, riemanli (Rieman) wrote…</i></summary><blockquote>

Could we ask them not to migrate the custom direct methodology before the next release?
</blockquote></details>

We could, but we could also just not rename the field. 

I think the idea is to get everyone to migrate to this before the next release so we can turn on variance.

**Review by kungfucraig — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/178)* status: all files reviewed, 3 unresolved discussions (waiting on @jiayu-google, @riemanli, @SanjayVas, and @stevenwarejones)

___
*[`src/main/proto/wfa/measurement/api/v2alpha/direct_computation.proto` line 42 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/178#-NfSVosB1DiPIP0tJdyX:-NfSqWOrCUM-pdGBnI2z:btil6tz) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/311d0595d7bd2d3956232a960d6dac358ef1ea9c/src/main/proto/wfa/measurement/api/v2alpha/direct_computation.proto#L42)):*
<details><summary><i>Previously, kungfucraig (Craig Wright) wrote…</i></summary><blockquote>

We could, but we could also just not rename the field. 

I think the idea is to get everyone to migrate to this before the next release so we can turn on variance.
</blockquote></details>

After some discussion with Rieman offline I'm starting to come around to his opinion. @SanjayVas WDYT?

**Review by stevenwarejones — CHANGES_REQUESTED** [review]
> Reviewed 1 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/178)* status: all files reviewed, 3 unresolved discussions (waiting on @jiayu-google, @riemanli, and @SanjayVas)

___
*[`src/main/proto/wfa/measurement/api/v2alpha/protocol_config.proto` line 145 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/178#-NfSRyAnELI_eETN54X2:-NfWVyyvExotUQr8Tinu:b-wgcgsz) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/311d0595d7bd2d3956232a960d6dac358ef1ea9c/src/main/proto/wfa/measurement/api/v2alpha/protocol_config.proto#L145)):*
<details><summary><i>Previously, kungfucraig (Craig Wright) wrote…</i></summary><blockquote>

Can we say "This guidance only applies to direct measurements, not MPC measurements."

I think eventually I'd like the Kingdom to generate a distinct salt for each MPC measurement. This would help ensure that if anyone got a hold of raw MPC data the ability to use it and combine it with other data would be limited. So, mostly just objecting to including the word "never."
</blockquote></details>

lets say someone used a custom methodology out-of-the-box like hll++ https://docs.databricks.com/en/sql/language-manual/functions/approx_count_distinct.html -- I think we'd need to provide better guidance how to interact with this API

**Review by kungfucraig — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/178)* status: all files reviewed, 3 unresolved discussions (waiting on @jiayu-google, @riemanli, @SanjayVas, and @stevenwarejones)

___
*[`src/main/proto/wfa/measurement/api/v2alpha/protocol_config.proto` line 145 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/178#-NfSRyAnELI_eETN54X2:-NfWfEBS6dPf3VsDqqy-:b-o9lgb) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/311d0595d7bd2d3956232a960d6dac358ef1ea9c/src/main/proto/wfa/measurement/api/v2alpha/protocol_config.proto#L145)):*
<details><summary><i>Previously, stevenwarejones (Steven Ware Jones) wrote…</i></summary><blockquote>

lets say someone used a custom methodology out-of-the-box like hll++ https://docs.databricks.com/en/sql/language-manual/functions/approx_count_distinct.html -- I think we'd need to provide better guidance how to interact with this API
</blockquote></details>

What specifically would you add here?

**Review by kungfucraig — APPROVED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/178)* status: all files reviewed, 2 unresolved discussions (waiting on @jiayu-google, @riemanli, @SanjayVas, and @stevenwarejones)

**Review by SanjayVas — APPROVED** [review]
> Reviewed 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/178)* status: all files reviewed, 2 unresolved discussions (waiting on @jiayu-google, @riemanli, and @stevenwarejones)

___
*[`src/main/proto/wfa/measurement/api/v2alpha/direct_computation.proto` line 42 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/178#-NfSVosB1DiPIP0tJdyX:-NfY-cUKBNtJXPBimTgc:bijqehg) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/311d0595d7bd2d3956232a960d6dac358ef1ea9c/src/main/proto/wfa/measurement/api/v2alpha/direct_computation.proto#L42)):*
<details><summary><i>Previously, kungfucraig (Craig Wright) wrote…</i></summary><blockquote>

After some discussion with Rieman offline I'm starting to come around to his opinion. @SanjayVas WDYT?
</blockquote></details>

Renaming fields just breaks compilation, but not wire compatibility. They can migrate just fine now for scalar results and still be compatible with a future release.

**Review by stevenwarejones — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/178)* status: all files reviewed, 2 unresolved discussions (waiting on @jiayu-google and @riemanli)

___
*[`src/main/proto/wfa/measurement/api/v2alpha/protocol_config.proto` line 145 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/178#-NfSRyAnELI_eETN54X2:-NfkHW8o2NfAVNng5uoN:b-fr9hh3) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/311d0595d7bd2d3956232a960d6dac358ef1ea9c/src/main/proto/wfa/measurement/api/v2alpha/protocol_config.proto#L145)):*
<details><summary><i>Previously, kungfucraig (Craig Wright) wrote…</i></summary><blockquote>

What specifically would you add here?
</blockquote></details>

maybe just use more common language like 
```
// salting their hash function / providing a random seed.
```

**Comment by riemanli** [comment]
*[`src/main/proto/wfa/measurement/api/v2alpha/protocol_config.proto` line 145 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/178#-NfSRyAnELI_eETN54X2:-NflTptO3J6uci1rFMQi:bhug1ke) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/311d0595d7bd2d3956232a960d6dac358ef1ea9c/src/main/proto/wfa/measurement/api/v2alpha/protocol_config.proto#L145)):*
<details><summary><i>Previously, stevenwarejones (Steven Ware Jones) wrote…</i></summary><blockquote>

maybe just use more common language like 
```
// salting their hash function / providing a random seed.
```

</blockquote></details>

"providing a random seed" sounds to me a bit too broad and ambiguous. Could you elaborate what that means? Does it have the same effect as what salting hash functions does?

**Comment by riemanli** [comment]
*[`src/main/proto/wfa/measurement/api/v2alpha/protocol_config.proto` line 145 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/178#-NfSRyAnELI_eETN54X2:-NflZLY4-ezmlSg4ANOn:b-dvjlwl) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/311d0595d7bd2d3956232a960d6dac358ef1ea9c/src/main/proto/wfa/measurement/api/v2alpha/protocol_config.proto#L145)):*
<details><summary><i>Previously, riemanli (Rieman) wrote…</i></summary><blockquote>

"providing a random seed" sounds to me a bit too broad and ambiguous. Could you elaborate what that means? Does it have the same effect as what salting hash functions does?
</blockquote></details>

Would "inserting a random seed into the hash function" work?

**Review by stevenwarejones — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/178)* status: all files reviewed, 2 unresolved discussions (waiting on @jiayu-google and @riemanli)

___
*[`src/main/proto/wfa/measurement/api/v2alpha/protocol_config.proto` line 145 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/178#-NfSRyAnELI_eETN54X2:-NfmA6DF29gOFTQDNzcs:b-3ej2ae) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/311d0595d7bd2d3956232a960d6dac358ef1ea9c/src/main/proto/wfa/measurement/api/v2alpha/protocol_config.proto#L145)):*
<details><summary><i>Previously, riemanli (Rieman) wrote…</i></summary><blockquote>

Would "inserting a random seed into the hash function" work?
</blockquote></details>

yes - I believe we should just call it a seed here and not salt, since it need not be kept secret?

**Review by riemanli — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/178)* status: all files reviewed, 1 unresolved discussion (waiting on @jiayu-google and @stevenwarejones)

___
*[`src/main/proto/wfa/measurement/api/v2alpha/protocol_config.proto` line 144 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/178#-NfSSFqL4KNxn36ARov2:-NfSY_ZdEmMzUMceQbpK:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/311d0595d7bd2d3956232a960d6dac358ef1ea9c/src/main/proto/wfa/measurement/api/v2alpha/protocol_config.proto#L144)):*
<details><summary><i>Previously, jiayu-google wrote…</i></summary><blockquote>

nit: use this methodology -> use any custom direct methodology
</blockquote></details>

Done.
___
*[`src/main/proto/wfa/measurement/api/v2alpha/protocol_config.proto` line 145 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/178#-NfSRyAnELI_eETN54X2:-NfmFaQN52nRIdEfp1N_:b-l5var7) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/311d0595d7bd2d3956232a960d6dac358ef1ea9c/src/main/proto/wfa/measurement/api/v2alpha/protocol_config.proto#L145)):*
<details><summary><i>Previously, stevenwarejones (Steven Ware Jones) wrote…</i></summary><blockquote>

yes - I believe we should just call it a seed here and not salt, since it need not be kept secret?
</blockquote></details>

Sounds good to me. @jiayu-google @kungfucraig  you agree?

**Review by jiayu-google — APPROVED** [review]
:lgtm:

> Reviewed 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/178#-:-NfmG1fV1m9XWgKL3bJY:bnfp4nl)* status: all files reviewed, 1 unresolved discussion (waiting on @stevenwarejones)

___
*[`src/main/proto/wfa/measurement/api/v2alpha/protocol_config.proto` line 145 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/178#-NfSRyAnELI_eETN54X2:-NfmFukj1vwn_qVXcTCP:b-81nlyc) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/311d0595d7bd2d3956232a960d6dac358ef1ea9c/src/main/proto/wfa/measurement/api/v2alpha/protocol_config.proto#L145)):*
<details><summary><i>Previously, riemanli (Rieman) wrote…</i></summary><blockquote>

Sounds good to me. @jiayu-google @kungfucraig  you agree?
</blockquote></details>

SGTM.

**Review by riemanli — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/178)* status: 0 of 2 files reviewed, 1 unresolved discussion (waiting on @jiayu-google, @kungfucraig, @SanjayVas, and @stevenwarejones)

___
*[`src/main/proto/wfa/measurement/api/v2alpha/protocol_config.proto` line 145 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/178#-NfSRyAnELI_eETN54X2:-NfmHAlk-RCGkeho3Ktj:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/311d0595d7bd2d3956232a960d6dac358ef1ea9c/src/main/proto/wfa/measurement/api/v2alpha/protocol_config.proto#L145)):*
<details><summary><i>Previously, jiayu-google wrote…</i></summary><blockquote>

SGTM.
</blockquote></details>

Done.

**Review by SanjayVas — APPROVED** [review]
> Reviewed 1 of 2 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/178)* status: all files reviewed, 1 unresolved discussion (waiting on @kungfucraig and @stevenwarejones)

**Review by kungfucraig — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/178)* status: all files reviewed, 1 unresolved discussion (waiting on @riemanli)

___
*[`src/main/proto/wfa/measurement/api/v2alpha/protocol_config.proto` line 145 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/178#-NfSRyAnELI_eETN54X2:-Nfq9X_f8Nl6Y3WKo_ZC:b-yasg6d) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/311d0595d7bd2d3956232a960d6dac358ef1ea9c/src/main/proto/wfa/measurement/api/v2alpha/protocol_config.proto#L145)):*
<details><summary><i>Previously, riemanli (Rieman) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

s/inserting a random seed/generating a random seed that is unique for the particular measurement and seeding their hash function with it.

**Review by kungfucraig — APPROVED** [review]
> Reviewed 1 of 2 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/178)* status: all files reviewed, 1 unresolved discussion (waiting on @riemanli)

**Review by riemanli — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/178)* status: 0 of 2 files reviewed, 1 unresolved discussion (waiting on @kungfucraig, @SanjayVas, and @stevenwarejones)

___
*[`src/main/proto/wfa/measurement/api/v2alpha/direct_computation.proto` line 32 at r4](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/178#-NfpxQ5E1dt1ynErNiB-:-Nfq_mws3cNuEbYidhjs:bf0lqm4) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/05b55aee1fd3dd18a774f96ae21387fac0638764/src/main/proto/wfa/measurement/api/v2alpha/direct_computation.proto#L32)):*
<details><summary><i>Previously, stevenwarejones (Steven Ware Jones) wrote…</i></summary><blockquote>

I was a little confused initially what you meant by a k+ frequency distribution. Can you define it here?
</blockquote></details>

Thanks for the suggestion. Added.
___
*[`src/main/proto/wfa/measurement/api/v2alpha/protocol_config.proto` line 145 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/178#-NfSRyAnELI_eETN54X2:-NfqYLDS2Rb0MPvOjscN:b-xu0uci) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/311d0595d7bd2d3956232a960d6dac358ef1ea9c/src/main/proto/wfa/measurement/api/v2alpha/protocol_config.proto#L145)):*
<details><summary><i>Previously, kungfucraig (Craig Wright) wrote…</i></summary><blockquote>

s/inserting a random seed/generating a random seed that is unique for the particular measurement and seeding their hash function with it.
</blockquote></details>

Thanks for the suggestion. Done.

**Review by kungfucraig — APPROVED** [review]
> Reviewed 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/178)* status: all files reviewed, 1 unresolved discussion (waiting on @SanjayVas and @stevenwarejones)

**Review by stevenwarejones — CHANGES_REQUESTED** [review]
> Reviewed 1 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/178)* status: all files reviewed, 1 unresolved discussion (waiting on @riemanli and @SanjayVas)

___
*[`src/main/proto/wfa/measurement/api/v2alpha/direct_computation.proto` line 26 at r5](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/178#-NfrSBeF-KekT4RRqK-B:-NfrSBeF-KekT4RRqK-C:brwemuk) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/4e8bbd0dc55656df56356d9129c928f72cfdf82a/src/main/proto/wfa/measurement/api/v2alpha/direct_computation.proto#L26)):*
> ```proto
> 
> // Information about the custom direct methodology.
> message CustomDirectMethodology {
> ```

shouldn't we actually split this into two messages: `CustomDirectMethodology` which will just have a scalar and a `CustomDirectHistogramMethology` that contains FrequencyVariances?

**Review by kungfucraig — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/178)* status: all files reviewed, 1 unresolved discussion (waiting on @SanjayVas and @stevenwarejones)

___
*[`src/main/proto/wfa/measurement/api/v2alpha/direct_computation.proto` line 26 at r5](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/178#-NfrSBeF-KekT4RRqK-B:-NfrTowR97DCoBXpvTDK:b-4fz6sd) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/4e8bbd0dc55656df56356d9129c928f72cfdf82a/src/main/proto/wfa/measurement/api/v2alpha/direct_computation.proto#L26)):*
<details><summary><i>Previously, stevenwarejones (Steven Ware Jones) wrote…</i></summary><blockquote>

shouldn't we actually split this into two messages: `CustomDirectMethodology` which will just have a scalar and a `CustomDirectHistogramMethology` that contains FrequencyVariances?
</blockquote></details>

We could do it that way. If so we have a consideration:

1. Remove CustomDirect from frequency and cause a binary and source incompatibility.
2. Leave CustomDirect on frequency and cause no incompatibility, but any existing clients using CustomDirect on frequency will not be providing the correct value.

The current PR maintains binary compatibility - we won't break running code. But it breaks source code compatibility, which means it will eventually force EDPs to update.

I don't like #2 above at all, and #1 is marginally better or worse than what we are doing depending on one's judgement. 

If we could go with what we have here, we can resolve this and move on to other things sooner. :)
